sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -788,12 +788,37 @@ object ScalaReflection extends ScalaReflection {
   }
 
   /**
-   * Finds an accessible constructor with compatible parameters. This is a more flexible search
-   * than the exact matching algorithm in `Class.getConstructor`. The first assignment-compatible
-   * matching constructor is returned. Otherwise, it returns `None`.
+   * Finds an accessible constructor with compatible parameters. This is a more flexible search than
+   * the exact matching algorithm in `Class.getConstructor`. The first assignment-compatible
+   * matching constructor is returned if it exists. Otherwise, we check for additional compatible
+   * constructors defined in the companion object as `apply` methods. Otherwise, it returns `None`.
    */
-  def findConstructor(cls: Class[_], paramTypes: Seq[Class[_]]): Option[Constructor[_]] = {
-    Option(ConstructorUtils.getMatchingAccessibleConstructor(cls, paramTypes: _*))
+  def findConstructor(cls: Class[_], paramTypes: Seq[Class[_]]): Option[Seq[AnyRef] => Any] = {
```

Comment:
Wouldn't the result also be AnyRef, if this produces an Object?
Is it not possible to write this with a type T for the class and its result? maybe there's a reason.

Reply:
I think you are correct, but I was keeping with the existing return type used in where this is called in `eval` and elsewhere and don't know if I should make changes not related to my issue. Same thing with changing `_` to `T`. Would require adding a type parameter to `NewInstance` which would be a very large change.

Reply:
A Constructor can only be for an Object, so it was already only returning something that can make any AnyRef. I think that's better.

Reply:
Real fast I get:
```
[error] /home/drew/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:798: type mismatch;
[error]  found   : _$31 where type _$31
[error]  required: T
[error]       case Some(c) => Some((x: Seq[AnyRef]) => c.newInstance(x: _*))
[error]                                                             ^
[error] /home/drew/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:818: type mismatch;
[error]  found   : Any
[error]  required: T
[error]             method.apply(args: _*)
[error]  
```

`Constructor` has a type parameter, but  the signature of `newInstance` doesn't use it, and in the second case `method.apply` method returned by reflection has a return type of `Any`.  I think that can only be resolved with adding `asInstanceOf[Seq[AnyRef] => T]`, which I don't think would be any better (unless I'm missing something)?

Reply:
I guess externally `findConstructor` would then be cleaner though? It's too bad it is only called in one place.

Reply:
Ok, that makes total sense. I've made this change also